pkg/database/connection.go

```diff
@@ -6,51 +6,65 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/driftprogramming/pgxpoolmock"
 	pgxpool "github.com/jackc/pgx/v4/pgxpool"
-	"github.com/open-cluster-management/search-indexer/pkg/config"
+	"github.com/stolostron/search-indexer/pkg/config"
 	"k8s.io/klog/v2"
 )
 
-var pool *pgxpool.Pool
+// Database Access Object. Use a DAO instance so we can replace the pool object in the unit tests.
+type DAO struct {
+	pool      pgxpoolmock.PgxPool
+	batchSize int
+}
+
+var poolSingleton pgxpoolmock.PgxPool
+
+// Creates new DAO instance.
+func NewDAO(p pgxpoolmock.PgxPool) DAO {
+	// Crete DAO with default values.
+	dao := DAO{
+		batchSize: 500,
+	}
+	if p != nil {
+		klog.Error("Using provided database connection. This path should only get executed during unit tests.")
+		dao.pool = p
+		return dao
+	}
 
-func init() {
-	klog.Info("Initializing database connection.")
-	initializePool()
+	if poolSingleton == nil {
+		poolSingleton = initializePool()
+	}
+	dao.pool = poolSingleton
+	return dao
 }
 
-func initializePool() {
-	cfg := config.New()
+func initializePool() pgxpoolmock.PgxPool {
+	cfg := config.Cfg
 
 	database_url := fmt.Sprintf("postgresql://%s:%s@%s:%d/%s", cfg.DBUser, cfg.DBPass, cfg.DBHost, cfg.DBPort, cfg.DBName)
 	klog.Info("Connecting to PostgreSQL at: ", fmt.Sprintf("postgresql://%s:%s@%s:%d/%s", cfg.DBUser, "*****", cfg.DBHost, cfg.DBPort, cfg.DBName))
 
 	config, configErr := pgxpool.ParseConfig(database_url)
 	if configErr != nil {
-		klog.Error("Error parsing database connection configuration.", configErr)
+		klog.Fatal("Error parsing database connection configuration. ", configErr)
 	}
 
 	conn, err := pgxpool.ConnectConfig(context.Background(), config)
 	if err != nil {
 		klog.Error("Unable to connect to database: %+v\n", err)
+		// TODO: We need to retry the connection until successful.
 	}
 
-	pool = conn
+	return conn
 }
 
-func GetConnection() *pgxpool.Pool {
-	if pool == nil {
-		initializePool()
-	}
+func (dao *DAO) InitializeTables() {
+	// FIXME: REMOVE THIS WORKAROUND! Dropping tables to simplify development, we can't keep this for production.
+	klog.Warning("FIXME: REMOVE THIS WORKAROUND! I'm dropping tables to simplify development, we can't keep this for production.")
 
-	if pool != nil {
-		err := pool.Ping(context.Background())
-		if err != nil {
-			klog.Error("Unable to get a database connection. ", err)
-			// Here we may need to add retry.
-			return nil
-		}
-		klog.Info("Successfully connected to database!")
-		return pool
-	}
-	return nil
+	dao.pool.Exec(context.Background(), "DROP TABLE resources")
+	dao.pool.Exec(context.Background(), "DROP TABLE edges")
+	dao.pool.Exec(context.Background(), "CREATE TABLE IF NOT EXISTS resources (uid TEXT PRIMARY KEY, cluster TEXT, data JSONB)")
```

Comment:
We should create a schema `search` and create tables in there.

Reply:
Agree, will add in next PR. Do we have a sample in our prototypes?